.github/workflows/build_wheels.yml

```diff
@@ -269,7 +269,7 @@ jobs:
         # TODO: https://github.com/apache/beam/issues/23048
         CIBW_SKIP: "*-musllinux_*"
         CIBW_ENVIRONMENT: "SETUPTOOLS_USE_DISTUTILS=stdlib"
-        CIBW_BEFORE_BUILD: pip install cython==0.29.36 numpy && pip install --upgrade setuptools
+        CIBW_BEFORE_BUILD: pip install cython==3.0.0 numpy && pip install --upgrade setuptools
```

Comment:
This line is what will cause most change to users: we will now build beam wheels with Cython 3.0.

1) Honestly, I would sit on Cython 3.0 for a bit and see what issues are reported, if any. I don't mind using Cython 3.0 in tests asap. 

2) I suggest we merge this as soon as 2.50.0 is cut, that would give us ~6 weeks buffer time. Let's cut an issue blocking 2.51.0 to revisit that we are still ok with shipping wheels built on Cython 3. Alternative: submit now, but revert this for the 2.50.0 release branch.

3) Let's announce in CHANGES.md: Beam wheels are now built with Cython 3.0.

Reply:
Sounds reasonable to me. We can get the fixes in now rather than later though.

Reply:
Do we need to `s/nogil/noexcept nogil/` in Beam codebase? I think there is one file that uses these directives. I came across reports of performance regression associated with this, e.g https://github.com/benfred/implicit/issues/678.

See also: https://cython.readthedocs.io/en/latest/src/userguide/migrating_to_cy30.html

Reply:
@robertwb to comment

Reply:
That instance should not have to be changed. We should do an inventory on our non-extern cdef functions (mostly declared in pxd files) and see if they're performance critical and never return exceptions and mark them as such.

Reply:
Makes sense. https://cython.readthedocs.io/en/latest/src/userguide/migrating_to_cy30.html#exception-values-and-noexcept explains the change and rationale, and also calls out the performance pitfall for `nogil` case.